include/deal.II/matrix_free/tools.h

```diff
@@ -442,31 +442,47 @@ namespace MatrixFreeTools
     public:
       static const unsigned int n_lanes = VectorizedArrayType::size();
 
-      ComputeDiagonalHelper(FEEvaluation<dim,
-                                         fe_degree,
-                                         n_q_points_1d,
-                                         n_components,
-                                         Number,
-                                         VectorizedArrayType> &phi)
-        : phi(phi)
+      ComputeDiagonalHelper()
+        : phi(nullptr)
       {}
 
+      ComputeDiagonalHelper(const ComputeDiagonalHelper &)
+        : phi(nullptr)
+      {}
+
+      void
+      initialize(FEEvaluation<dim,
+                              fe_degree,
+                              n_q_points_1d,
+                              n_components,
+                              Number,
+                              VectorizedArrayType> &phi)
+      {
+        // if we are in hp mode and the number of unknowns changed, we must
+        // clear the map of entries
+        if (this->phi != nullptr &&
+            this->phi->dofs_per_component != phi.dofs_per_component)
+          locally_relevant_constraints_hn_map.clear();
```

Comment:
:+1:  Looks good. In practice, the fast hn algorithm is disabled if hp is detected.